docs/design/sql/02-operator-interface.md

```diff
@@ -0,0 +1,157 @@
+
+An example of a query, its parse tree, and its relational tree is provided below.
+
+*Snippet 1: Query*
+```sql
+SELECT a, SUM(b)
+FROM table
+GROUP BY a
+HAVING SUM(b) > 50
+```
+*Snippet 2: Parse Tree*
+```
+-- Select
+---- SelectList [a, SUM(b)]
+---- From [table]
+---- GroupBy [a]
+---- Having [SUM(b) > 50]
+```
+*Snippet 3: Relational Tree*
+```
+-- Filter [SUM(b) > 50]
+---- Aggregate [a -> SUM(b)]
+------ Project [a, b]
+-------- Scan [table]
+```
+
+## Volcano Model
+
+Volcano Model defines the common data exchange interface between operators in the relational tree. This allows
+for extensibility, as new operators could be implemented with minimal changes to the engine.
+
+In the original paper the interface consists of three operations:
+
+*Snippet 4: Volcano Interface*
+```java
+interface Operator {
+    void open();  // Initialize the operator
+    Row next();   // Get the next row
+    void close(); // Close the operator and release all resources
+}
+```
+
+## Volcano Model in Hazelcast Mustang
+
+The original Volcano Model has two drawbacks:
+1. Operators exchange one row at a time, which leads to performance overhead
+2. Call to the `next()` is blocking, which is not optimal for the distributed environment, where
+operators often wait for remote data or free space in the send buffer.
+
+To achieve high performance, we introduce several changes to the original Volcano Model: batching and
+non-blocking execution.
+
+### Row and RowBatch
+We define the `RowBatch` interface which a collection of rows (tuples).
+
+*Snippet 5: RowBatch interface*
+```java
+interface RowBatch {
+    Row getRow(int index); // Get the row by index
+    int getRowCount();     // Get the number of rows 
+} 
+```
+
+Then we define the `Row` interface, which provides access to values by index. The `Row` itself is considered
+as a special case of `RowBatch` with one row. This allows saving on allocations in some parts of the engine.
+
+*Snippet 6: Row interface*
+```java
+interface Row extends RowBatch {
+    Object get(int index); // Get the value by index
+    int getColumnCount();  // Get the number of values in the row 
+    
+    default int getRowCount() {
+        return 1;
+    }
+    
+    default int getRow(int index) {
+        return this;
+    }
+}
+```
+
+### Operator
+The operator is defined by `Exec` interface:
+1. Operators exchange `RowBatch` instead of `Row`
+1. The blocking `next()` method is replaced with the non-blocking `advance()` method, which returns the iteration
+result instead of the row batch
+1. The `RowBatch` could be accessed through a separate method
+1. The `open()` method is renamed to `setup()`. Special query context is passed to it as an argument
+1. There is no separate `close()` method because the engine doesn't need explicit per-operator cleanup at the
+moment. This may change in future, in this case the current document should be updated accordingly
+
+*Snippet 7: Exec interface*
+```java
+interface Exec {
```

Comment:
If it's referred everywhere as an operator, why not just name it the same way?

Reply:
The operator is an ambiguous term. Its precise meaning depends on the context. Calcite nodes (`RelNode`, `LogicalRel`, `PhysicalRel`), plan nodes (`PhysicalNode`), and iterator nodes (`Exec`) are all operators. So in different design documents, we will refer to different entities as "operator". 
`Exec` is used here for historical reasons, to stress out that this is an **exec**utable node, as opposed to relational operators used during optimization. Moreover, certain executors, such as `SendExec` or `ReceiveExec`, have no relational operator counterparts at all, so `Exec` is a slightly more broad term than `relational operator`.
I slightly changed the name of the snippet itself, but I would avoid changing interface name for now since it is an internal thing which is unlikely to be touched by anyone except for advanced product hackers, but at the same time renaming might cause conflicts between multiple prototype and implementation branches.